src/diagram/components/ui/expandable-input.tsx

```diff
@@ -84,16 +98,20 @@ export const ExpandableInput = ({
     }
   };
 
+  // Ensure that hasLongValue is updated when the value changes in 
+  // disabled fields -- the onChange handler above does not get triggered
+  // by disabled fields.
+  useEffect(() => {
+    if (disabled && value) {
+      setHasLongValue(isLongValue(value, lengthToExpand));
+    }
+  }, [disabled, lengthToExpand, value]);
+
```

Comment:
Good insight about the `useEffect`, that makes sense why it would be necessary.

Your question about `onChange` vs `onBlur` in `NumberInput` is a great one, and unfortunately I don't have a great answer. The main reason I remember using `onBlur` rather than `onChange` for the `NumberInput` is that's when PIs requested the value change. I can think of some possible technical reasons to not to it, but I'm not sure how legitimate those reasons are. So I think it would be a good idea to experiment with switching when `NumberInput` updates its value. If you do, you should test the change in the new/edit variable dialogs as well as the variable card.

Reply:
This is basically the distinction between controlled (`onChange`) vs. uncontrolled (`onBlur`) components. If you don't need external access to the value between focus and blur, then uncontrolled is generally simpler. If you _do_ need external access to the value, then a controlled component is generally the solution. You're essentially creating a semi-controlled component here by not putting the value in state but then putting `hasLongValue` in state. I suspect you could eliminate `hasLongValue` from state by having the value in state instead, thus switching to a controlled component. Not sure whether it's worth it, though, so I leave that up to you. 😉